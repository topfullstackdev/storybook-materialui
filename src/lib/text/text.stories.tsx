import { Meta, Story } from '@storybook/react';
import { ElementType } from 'react';

import { Text, TextProps } from './text';

export default {
  title: 'Text',
  component: Text,
} as Meta;

const Template: Story<TextProps<ElementType>> = function Template(args) {
  // eslint-disable-next-line @typescript-eslint/no-unsafe-assignment
  const { children, ...rest } = args;
  return <Text {...rest}>{children}</Text>;
};

export const Primary = Template.bind({});
Primary.args = {
  children: 'Text',
  variant: 'body1',
  italicized: false,
  bold: false,
};

const VariantsTemplate: Story<TextProps<ElementType>> =
  function VariantsTemplate(args) {
    const variants = [
      'h1',
      'h2',
      'h3',
      'h4',
      'h5',
      'h6',
      'body1',
      'body2',
      'subtitle1',
      'subtitle2',
      'overline',
      'caption',
      'button',
      'inherit',
    ] as const;
    return (
      <>
        {variants.map((variant) => (
          <Text variant={variant} key={variant} display="block" {...args}>
            {variant}
          </Text>
        ))}
      </>
    );
  };

export const Variants = VariantsTemplate.bind({});
