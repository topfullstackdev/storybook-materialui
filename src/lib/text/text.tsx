import styled from '@emotion/styled';
import Typography from '@mui/material/Typography';

import type { TypographyProps } from '@mui/material';

export type TextProps<C extends React.ElementType> = TypographyProps<
  C,
  { component?: C }
> & {
  bold?: boolean;
  italicized?: boolean;
  underlined?: boolean;
  children?: React.ReactNode;
};

const TextComponent = styled(Typography, {
  shouldForwardProp: (prop: string) =>
    !['bold', 'italicized', 'underlined'].includes(prop),
})<TextProps<React.ElementType>>`
  font-weight: ${({ bold }) => (bold ? 600 : 'normal')};
  font-style: ${({ italicized }) => (italicized ? 'italic' : 'normal')};
  text-decoration: ${({ underlined }) => (underlined ? 'underline' : 'none')};
  font-family: ${({ variant }) =>
    variant?.startsWith('h') && variant !== 'h6'
      ? 'Rockwell,Verdana,Arial,san-serif'
      : 'Verdana,Arial,sans-serif'};
` as typeof Typography;

export function Text<C extends React.ElementType>({
  children,
  ...props
}: TextProps<C>) {
  return <TextComponent {...props}>{children}</TextComponent>;
}
