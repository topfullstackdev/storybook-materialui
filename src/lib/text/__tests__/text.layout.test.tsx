import { render } from '@testing-library/react';
import { flatten } from 'lodash';

import { Text } from '../text';

describe('Text', () => {
  const variants = [
    'h1',
    'h2',
    'h3',
    'h4',
    'h5',
    'h6',
    'body1',
    'body2',
    'subtitle1',
    'subtitle2',
    'overline',
    'caption',
    'button',
    'inherit',
  ] as const;
  const matrix: [typeof variants[number], boolean, boolean, boolean][] =
    flatten(
      variants.map((variant) => [
        [variant, true, true, true],
        [variant, false, false, false],
      ]),
    );
  it.each(matrix)(
    'Should render when variant=%s, italicized=%s, bold=%s, and underlined=%s',
    (variant, italicized, bold, underlined) => {
      const { container } = render(
        <Text
          variant={variant}
          italicized={italicized}
          bold={bold}
          underlined={underlined}
        >
          text
        </Text>,
      );

      expect(container).toMatchSnapshot();
    },
  );
});
