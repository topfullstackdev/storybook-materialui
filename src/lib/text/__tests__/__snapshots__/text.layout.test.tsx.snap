// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Text Should render when variant=body1, italicized=false, bold=false, and underlined=false 1`] = `
.emotion-1 {
  margin: 0;
  font-family: "Roboto","Helvetica","Arial",sans-serif;
  font-weight: 400;
  font-size: 1rem;
  line-height: 1.5;
  letter-spacing: 0.00938em;
  font-weight: normal;
  font-style: normal;
  -webkit-text-decoration: none;
  text-decoration: none;
  font-family: Verdana,Arial,sans-serif;
}

<div>
  <p
    class="MuiTypography-root MuiTypography-body1 emotion-0 emotion-1"
  >
    text
  </p>
</div>
`;

exports[`Text Should render when variant=body1, italicized=true, bold=true, and underlined=true 1`] = `
.emotion-1 {
  margin: 0;
  font-family: "Roboto","Helvetica","Arial",sans-serif;
  font-weight: 400;
  font-size: 1rem;
  line-height: 1.5;
  letter-spacing: 0.00938em;
  font-weight: 600;
  font-style: italic;
  -webkit-text-decoration: underline;
  text-decoration: underline;
  font-family: Verdana,Arial,sans-serif;
}

<div>
  <p
    class="MuiTypography-root MuiTypography-body1 emotion-0 emotion-1"
  >
    text
  </p>
</div>
`;

exports[`Text Should render when variant=body2, italicized=false, bold=false, and underlined=false 1`] = `
.emotion-1 {
  margin: 0;
  font-family: "Roboto","Helvetica","Arial",sans-serif;
  font-weight: 400;
  font-size: 0.875rem;
  line-height: 1.43;
  letter-spacing: 0.01071em;
  font-weight: normal;
  font-style: normal;
  -webkit-text-decoration: none;
  text-decoration: none;
  font-family: Verdana,Arial,sans-serif;
}

<div>
  <p
    class="MuiTypography-root MuiTypography-body2 emotion-0 emotion-1"
  >
    text
  </p>
</div>
`;

exports[`Text Should render when variant=body2, italicized=true, bold=true, and underlined=true 1`] = `
.emotion-1 {
  margin: 0;
  font-family: "Roboto","Helvetica","Arial",sans-serif;
  font-weight: 400;
  font-size: 0.875rem;
  line-height: 1.43;
  letter-spacing: 0.01071em;
  font-weight: 600;
  font-style: italic;
  -webkit-text-decoration: underline;
  text-decoration: underline;
  font-family: Verdana,Arial,sans-serif;
}

<div>
  <p
    class="MuiTypography-root MuiTypography-body2 emotion-0 emotion-1"
  >
    text
  </p>
</div>
`;

exports[`Text Should render when variant=button, italicized=false, bold=false, and underlined=false 1`] = `
.emotion-1 {
  margin: 0;
  font-family: "Roboto","Helvetica","Arial",sans-serif;
  font-weight: 500;
  font-size: 0.875rem;
  line-height: 1.75;
  letter-spacing: 0.02857em;
  text-transform: uppercase;
  font-weight: normal;
  font-style: normal;
  -webkit-text-decoration: none;
  text-decoration: none;
  font-family: Verdana,Arial,sans-serif;
}

<div>
  <span
    class="MuiTypography-root MuiTypography-button emotion-0 emotion-1"
  >
    text
  </span>
</div>
`;

exports[`Text Should render when variant=button, italicized=true, bold=true, and underlined=true 1`] = `
.emotion-1 {
  margin: 0;
  font-family: "Roboto","Helvetica","Arial",sans-serif;
  font-weight: 500;
  font-size: 0.875rem;
  line-height: 1.75;
  letter-spacing: 0.02857em;
  text-transform: uppercase;
  font-weight: 600;
  font-style: italic;
  -webkit-text-decoration: underline;
  text-decoration: underline;
  font-family: Verdana,Arial,sans-serif;
}

<div>
  <span
    class="MuiTypography-root MuiTypography-button emotion-0 emotion-1"
  >
    text
  </span>
</div>
`;

exports[`Text Should render when variant=caption, italicized=false, bold=false, and underlined=false 1`] = `
.emotion-1 {
  margin: 0;
  font-family: "Roboto","Helvetica","Arial",sans-serif;
  font-weight: 400;
  font-size: 0.75rem;
  line-height: 1.66;
  letter-spacing: 0.03333em;
  font-weight: normal;
  font-style: normal;
  -webkit-text-decoration: none;
  text-decoration: none;
  font-family: Verdana,Arial,sans-serif;
}

<div>
  <span
    class="MuiTypography-root MuiTypography-caption emotion-0 emotion-1"
  >
    text
  </span>
</div>
`;

exports[`Text Should render when variant=caption, italicized=true, bold=true, and underlined=true 1`] = `
.emotion-1 {
  margin: 0;
  font-family: "Roboto","Helvetica","Arial",sans-serif;
  font-weight: 400;
  font-size: 0.75rem;
  line-height: 1.66;
  letter-spacing: 0.03333em;
  font-weight: 600;
  font-style: italic;
  -webkit-text-decoration: underline;
  text-decoration: underline;
  font-family: Verdana,Arial,sans-serif;
}

<div>
  <span
    class="MuiTypography-root MuiTypography-caption emotion-0 emotion-1"
  >
    text
  </span>
</div>
`;

exports[`Text Should render when variant=h1, italicized=false, bold=false, and underlined=false 1`] = `
.emotion-1 {
  margin: 0;
  font-family: "Roboto","Helvetica","Arial",sans-serif;
  font-weight: 300;
  font-size: 6rem;
  line-height: 1.167;
  letter-spacing: -0.01562em;
  font-weight: normal;
  font-style: normal;
  -webkit-text-decoration: none;
  text-decoration: none;
  font-family: Rockwell,Verdana,Arial,san-serif;
}

<div>
  <h1
    class="MuiTypography-root MuiTypography-h1 emotion-0 emotion-1"
  >
    text
  </h1>
</div>
`;

exports[`Text Should render when variant=h1, italicized=true, bold=true, and underlined=true 1`] = `
.emotion-1 {
  margin: 0;
  font-family: "Roboto","Helvetica","Arial",sans-serif;
  font-weight: 300;
  font-size: 6rem;
  line-height: 1.167;
  letter-spacing: -0.01562em;
  font-weight: 600;
  font-style: italic;
  -webkit-text-decoration: underline;
  text-decoration: underline;
  font-family: Rockwell,Verdana,Arial,san-serif;
}

<div>
  <h1
    class="MuiTypography-root MuiTypography-h1 emotion-0 emotion-1"
  >
    text
  </h1>
</div>
`;

exports[`Text Should render when variant=h2, italicized=false, bold=false, and underlined=false 1`] = `
.emotion-1 {
  margin: 0;
  font-family: "Roboto","Helvetica","Arial",sans-serif;
  font-weight: 300;
  font-size: 3.75rem;
  line-height: 1.2;
  letter-spacing: -0.00833em;
  font-weight: normal;
  font-style: normal;
  -webkit-text-decoration: none;
  text-decoration: none;
  font-family: Rockwell,Verdana,Arial,san-serif;
}

<div>
  <h2
    class="MuiTypography-root MuiTypography-h2 emotion-0 emotion-1"
  >
    text
  </h2>
</div>
`;

exports[`Text Should render when variant=h2, italicized=true, bold=true, and underlined=true 1`] = `
.emotion-1 {
  margin: 0;
  font-family: "Roboto","Helvetica","Arial",sans-serif;
  font-weight: 300;
  font-size: 3.75rem;
  line-height: 1.2;
  letter-spacing: -0.00833em;
  font-weight: 600;
  font-style: italic;
  -webkit-text-decoration: underline;
  text-decoration: underline;
  font-family: Rockwell,Verdana,Arial,san-serif;
}

<div>
  <h2
    class="MuiTypography-root MuiTypography-h2 emotion-0 emotion-1"
  >
    text
  </h2>
</div>
`;

exports[`Text Should render when variant=h3, italicized=false, bold=false, and underlined=false 1`] = `
.emotion-1 {
  margin: 0;
  font-family: "Roboto","Helvetica","Arial",sans-serif;
  font-weight: 400;
  font-size: 3rem;
  line-height: 1.167;
  letter-spacing: 0em;
  font-weight: normal;
  font-style: normal;
  -webkit-text-decoration: none;
  text-decoration: none;
  font-family: Rockwell,Verdana,Arial,san-serif;
}

<div>
  <h3
    class="MuiTypography-root MuiTypography-h3 emotion-0 emotion-1"
  >
    text
  </h3>
</div>
`;

exports[`Text Should render when variant=h3, italicized=true, bold=true, and underlined=true 1`] = `
.emotion-1 {
  margin: 0;
  font-family: "Roboto","Helvetica","Arial",sans-serif;
  font-weight: 400;
  font-size: 3rem;
  line-height: 1.167;
  letter-spacing: 0em;
  font-weight: 600;
  font-style: italic;
  -webkit-text-decoration: underline;
  text-decoration: underline;
  font-family: Rockwell,Verdana,Arial,san-serif;
}

<div>
  <h3
    class="MuiTypography-root MuiTypography-h3 emotion-0 emotion-1"
  >
    text
  </h3>
</div>
`;

exports[`Text Should render when variant=h4, italicized=false, bold=false, and underlined=false 1`] = `
.emotion-1 {
  margin: 0;
  font-family: "Roboto","Helvetica","Arial",sans-serif;
  font-weight: 400;
  font-size: 2.125rem;
  line-height: 1.235;
  letter-spacing: 0.00735em;
  font-weight: normal;
  font-style: normal;
  -webkit-text-decoration: none;
  text-decoration: none;
  font-family: Rockwell,Verdana,Arial,san-serif;
}

<div>
  <h4
    class="MuiTypography-root MuiTypography-h4 emotion-0 emotion-1"
  >
    text
  </h4>
</div>
`;

exports[`Text Should render when variant=h4, italicized=true, bold=true, and underlined=true 1`] = `
.emotion-1 {
  margin: 0;
  font-family: "Roboto","Helvetica","Arial",sans-serif;
  font-weight: 400;
  font-size: 2.125rem;
  line-height: 1.235;
  letter-spacing: 0.00735em;
  font-weight: 600;
  font-style: italic;
  -webkit-text-decoration: underline;
  text-decoration: underline;
  font-family: Rockwell,Verdana,Arial,san-serif;
}

<div>
  <h4
    class="MuiTypography-root MuiTypography-h4 emotion-0 emotion-1"
  >
    text
  </h4>
</div>
`;

exports[`Text Should render when variant=h5, italicized=false, bold=false, and underlined=false 1`] = `
.emotion-1 {
  margin: 0;
  font-family: "Roboto","Helvetica","Arial",sans-serif;
  font-weight: 400;
  font-size: 1.5rem;
  line-height: 1.334;
  letter-spacing: 0em;
  font-weight: normal;
  font-style: normal;
  -webkit-text-decoration: none;
  text-decoration: none;
  font-family: Rockwell,Verdana,Arial,san-serif;
}

<div>
  <h5
    class="MuiTypography-root MuiTypography-h5 emotion-0 emotion-1"
  >
    text
  </h5>
</div>
`;

exports[`Text Should render when variant=h5, italicized=true, bold=true, and underlined=true 1`] = `
.emotion-1 {
  margin: 0;
  font-family: "Roboto","Helvetica","Arial",sans-serif;
  font-weight: 400;
  font-size: 1.5rem;
  line-height: 1.334;
  letter-spacing: 0em;
  font-weight: 600;
  font-style: italic;
  -webkit-text-decoration: underline;
  text-decoration: underline;
  font-family: Rockwell,Verdana,Arial,san-serif;
}

<div>
  <h5
    class="MuiTypography-root MuiTypography-h5 emotion-0 emotion-1"
  >
    text
  </h5>
</div>
`;

exports[`Text Should render when variant=h6, italicized=false, bold=false, and underlined=false 1`] = `
.emotion-1 {
  margin: 0;
  font-family: "Roboto","Helvetica","Arial",sans-serif;
  font-weight: 500;
  font-size: 1.25rem;
  line-height: 1.6;
  letter-spacing: 0.0075em;
  font-weight: normal;
  font-style: normal;
  -webkit-text-decoration: none;
  text-decoration: none;
  font-family: Verdana,Arial,sans-serif;
}

<div>
  <h6
    class="MuiTypography-root MuiTypography-h6 emotion-0 emotion-1"
  >
    text
  </h6>
</div>
`;

exports[`Text Should render when variant=h6, italicized=true, bold=true, and underlined=true 1`] = `
.emotion-1 {
  margin: 0;
  font-family: "Roboto","Helvetica","Arial",sans-serif;
  font-weight: 500;
  font-size: 1.25rem;
  line-height: 1.6;
  letter-spacing: 0.0075em;
  font-weight: 600;
  font-style: italic;
  -webkit-text-decoration: underline;
  text-decoration: underline;
  font-family: Verdana,Arial,sans-serif;
}

<div>
  <h6
    class="MuiTypography-root MuiTypography-h6 emotion-0 emotion-1"
  >
    text
  </h6>
</div>
`;

exports[`Text Should render when variant=inherit, italicized=false, bold=false, and underlined=false 1`] = `
.emotion-1 {
  margin: 0;
  font-weight: normal;
  font-style: normal;
  -webkit-text-decoration: none;
  text-decoration: none;
  font-family: Verdana,Arial,sans-serif;
}

<div>
  <p
    class="MuiTypography-root MuiTypography-inherit emotion-0 emotion-1"
  >
    text
  </p>
</div>
`;

exports[`Text Should render when variant=inherit, italicized=true, bold=true, and underlined=true 1`] = `
.emotion-1 {
  margin: 0;
  font-weight: 600;
  font-style: italic;
  -webkit-text-decoration: underline;
  text-decoration: underline;
  font-family: Verdana,Arial,sans-serif;
}

<div>
  <p
    class="MuiTypography-root MuiTypography-inherit emotion-0 emotion-1"
  >
    text
  </p>
</div>
`;

exports[`Text Should render when variant=overline, italicized=false, bold=false, and underlined=false 1`] = `
.emotion-1 {
  margin: 0;
  font-family: "Roboto","Helvetica","Arial",sans-serif;
  font-weight: 400;
  font-size: 0.75rem;
  line-height: 2.66;
  letter-spacing: 0.08333em;
  text-transform: uppercase;
  font-weight: normal;
  font-style: normal;
  -webkit-text-decoration: none;
  text-decoration: none;
  font-family: Verdana,Arial,sans-serif;
}

<div>
  <span
    class="MuiTypography-root MuiTypography-overline emotion-0 emotion-1"
  >
    text
  </span>
</div>
`;

exports[`Text Should render when variant=overline, italicized=true, bold=true, and underlined=true 1`] = `
.emotion-1 {
  margin: 0;
  font-family: "Roboto","Helvetica","Arial",sans-serif;
  font-weight: 400;
  font-size: 0.75rem;
  line-height: 2.66;
  letter-spacing: 0.08333em;
  text-transform: uppercase;
  font-weight: 600;
  font-style: italic;
  -webkit-text-decoration: underline;
  text-decoration: underline;
  font-family: Verdana,Arial,sans-serif;
}

<div>
  <span
    class="MuiTypography-root MuiTypography-overline emotion-0 emotion-1"
  >
    text
  </span>
</div>
`;

exports[`Text Should render when variant=subtitle1, italicized=false, bold=false, and underlined=false 1`] = `
.emotion-1 {
  margin: 0;
  font-family: "Roboto","Helvetica","Arial",sans-serif;
  font-weight: 400;
  font-size: 1rem;
  line-height: 1.75;
  letter-spacing: 0.00938em;
  font-weight: normal;
  font-style: normal;
  -webkit-text-decoration: none;
  text-decoration: none;
  font-family: Verdana,Arial,sans-serif;
}

<div>
  <h6
    class="MuiTypography-root MuiTypography-subtitle1 emotion-0 emotion-1"
  >
    text
  </h6>
</div>
`;

exports[`Text Should render when variant=subtitle1, italicized=true, bold=true, and underlined=true 1`] = `
.emotion-1 {
  margin: 0;
  font-family: "Roboto","Helvetica","Arial",sans-serif;
  font-weight: 400;
  font-size: 1rem;
  line-height: 1.75;
  letter-spacing: 0.00938em;
  font-weight: 600;
  font-style: italic;
  -webkit-text-decoration: underline;
  text-decoration: underline;
  font-family: Verdana,Arial,sans-serif;
}

<div>
  <h6
    class="MuiTypography-root MuiTypography-subtitle1 emotion-0 emotion-1"
  >
    text
  </h6>
</div>
`;

exports[`Text Should render when variant=subtitle2, italicized=false, bold=false, and underlined=false 1`] = `
.emotion-1 {
  margin: 0;
  font-family: "Roboto","Helvetica","Arial",sans-serif;
  font-weight: 500;
  font-size: 0.875rem;
  line-height: 1.57;
  letter-spacing: 0.00714em;
  font-weight: normal;
  font-style: normal;
  -webkit-text-decoration: none;
  text-decoration: none;
  font-family: Verdana,Arial,sans-serif;
}

<div>
  <h6
    class="MuiTypography-root MuiTypography-subtitle2 emotion-0 emotion-1"
  >
    text
  </h6>
</div>
`;

exports[`Text Should render when variant=subtitle2, italicized=true, bold=true, and underlined=true 1`] = `
.emotion-1 {
  margin: 0;
  font-family: "Roboto","Helvetica","Arial",sans-serif;
  font-weight: 500;
  font-size: 0.875rem;
  line-height: 1.57;
  letter-spacing: 0.00714em;
  font-weight: 600;
  font-style: italic;
  -webkit-text-decoration: underline;
  text-decoration: underline;
  font-family: Verdana,Arial,sans-serif;
}

<div>
  <h6
    class="MuiTypography-root MuiTypography-subtitle2 emotion-0 emotion-1"
  >
    text
  </h6>
</div>
`;
